src/Compiler/Microsoft.CodeAnalysis.Razor.Compiler/src/Language/Intermediate/IntermediateNodeWalker.Stack.cs

```diff
@@ -0,0 +1,75 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using System;
+using System.Diagnostics;
+using System.Diagnostics.CodeAnalysis;
+
+namespace Microsoft.AspNetCore.Razor.Language.Intermediate;
+
+public abstract partial class IntermediateNodeWalker
+{
+    /// <summary>
+    ///  Simple stack implementation that fills an array from end-to-start
+    ///  to keep the items in the reverse order they were pushed. This ensures
+    ///  that they are in the correct order to implement the <see cref="Ancestors"/>
+    ///  property.
+    /// </summary>
+    private struct Stack
+    {
+        private const int InitialStackSize = 4;
+
+        private IntermediateNode[]? _stack;
+        private int _stackPointer;
+
+        [MemberNotNullWhen(false, nameof(_stack))]
+        public readonly bool IsEmpty
+            => _stack is null || _stackPointer == _stack.Length;
+
+        public readonly ReadOnlySpan<IntermediateNode> Span
```

Comment:
>Span [](http://example.com/codeflow?start=55&length=4)

In stack nomenclature, wouldn't this be something like PeekAll()

Reply:
Maybe? This isn't a general data structure though, it's designed specifically for `InteredmateNodeWalker`. IMO, `Span` works a bit better for this particular purpose.

Reply:
OK, not a big deal to me